### PR TITLE
Feature dump-c-type-header flag

### PR DESCRIPTION
**Pull request by mmuesly**
I want to get the possibility for generating a header describing the internal types of a c module. This allows to access them in a harness without including the c file. Some project decide to define module local types that are only exported by typedefs in a header for the user. This doesn't allow to manipulate fields in the harness, if the harness is compiled without having the c module file in the compile context. Including the c file raises other problems regarding stubbing of functions. 
Therefore, I introduce the `--drop-c-type-header` flag to the `goto-instrument`  binary. This allows to create a header file than can be used to compile the harness. Later, the object files can be linked against the c module object file for CBMC verification.

@tautschnig suggested in a discussion during PR creation to eventually add this flag to the `goto-harness` binary. After careful thinking, I think this is not a good idea, because I don't see how the `dump-*` infrastructure harmonize with the generator driven approach of the goto-harness binary implementation. But I am open for a discussion, where the `--dump-c-type-header` flag should live instead of the `goto-instrument` binary. 

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-instrument/dump_c.cpp [hunk 1]
```diff
@@ -24,6 +24,36 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "dump_c_class.h"
 #include "goto_program2code.h"
 
+#include <iostream>
```
Comment (tautschnig) [inline src/goto-instrument/dump_c.cpp]: debugging left-over?

File: src/goto-instrument/dump_c.cpp [hunk 2]
```diff
@@ -282,35 +312,39 @@ void dump_ct::operator()(std::ostream &os)
        << "#endif\n\n";
   }
 
-  if(!global_decl_stream.str().empty())
-    os << global_decl_stream.str() << '\n';
+  if(!global_decl_stream.str().empty() &&
+    dump_c_config.include_global_decl)
+      os << global_decl_stream.str() << '\n';
```
Comment (tautschnig) [inline src/goto-instrument/dump_c.cpp]: I think clang-format will require some reformatting here, but also: please add braces around the body when either the condition or the body span multiple lines.

File: src/goto-instrument/dump_c.cpp [hunk 3]
```diff
@@ -24,6 +24,36 @@ Author: Daniel Kroening, kroening@kroening.com
...
+dump_c_configurationt dump_c_configurationt::default_configuration
+{
+  true,
+  true,
+  true,
+  true,
+  true,
+  true,
+  true,
+  false
```
Comment (tautschnig) [inline src/goto-instrument/dump_c.cpp]: Please add comments for each member.

File: src/goto-instrument/dump_c.cpp [hunk 4]
```diff
@@ -24,6 +24,36 @@ Author: Daniel Kroening, kroening@kroening.com
...
+dump_c_configurationt dump_c_configurationt::type_header_configuration
+{
+  false,
+  false,
+  true,
+  true,
+  true,
+  true,
+  true,
+  true
```
Comment (tautschnig) [inline src/goto-instrument/dump_c.cpp]: As above: comments, please.

File: src/goto-instrument/dump_c_class.h [hunk 5]
```diff
@@ -21,6 +21,63 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include <goto-programs/system_library_symbols.h>
 
+/// Used for configuring the behaviour of dump_c
+struct dump_c_configurationt final
+{
+  /// Include the function declaration in the dump
+  bool include_functions_decl;
```
Comment (tautschnig) [inline src/goto-instrument/dump_c_class.h]: Maybe `include_function_decls`?

File: src/goto-instrument/dump_c_class.h [hunk 6]
```diff
@@ -21,6 +21,63 @@ Author: Daniel Kroening, kroening@kroening.com
...
+
+  /// Include the functions in the dump
+  bool include_functions;
```
Comment (tautschnig) [inline src/goto-instrument/dump_c_class.h]: Maybe `include_function_bodies`?

File: src/goto-instrument/dump_c_class.h [hunk 7]
```diff
@@ -21,6 +21,63 @@ Author: Daniel Kroening, kroening@kroening.com
...
+  
+  /// Include the global declaration in the dump
+  bool include_global_decl;
```
Comment (tautschnig) [inline src/goto-instrument/dump_c_class.h]: Maybe `include_global_decls`? But also: what exactly is a "global declaration?"

File: scripts/extract_type_header.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env python3
```
Comment (tautschnig) [inline scripts/extract_type_header.py:1]: Random place: please review the commit message, it has a couple of typos: "required *to* access the *module*", "and *collect*"

File: scripts/extract_type_header.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+#
+# Dump type header for private types in c modules.
+#
+# Author: Malte Mues <mail.mues@gmail.com>
+# Copyright Amazon, Inc. 2019
```
Comment (tautschnig) [inline scripts/extract_type_header.py]: No need for the Amazon copyright here.

File: scripts/extract_type_header.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+                           header_file]
+        subprocess.run(drop_header_cmd,
+                       stdout=subprocess.PIPE,
+                       stderr=subprocess.STDOUT,
+                       check=True,
+                       universal_newlines=True,
+                       cwd=tmpdir,
+                       shell=False)
+
+        absolut_header = os.path.normpath(os.path.join(tmpdir, header_file))
```
Comment (tautschnig) [inline scripts/extract_type_header.py]: Nit pick: "absolut" isn't a part of the English language, how about simply using `header_path`?

File: scripts/extract_type_header.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+def get_module_name(c_file):
+    base = os.path.basename(c_file)
+    return base.split(".")[0]
+
+
+def make_header_file(goto_binary, c_file, header_out=None):
+    with TemporaryDirectory() as tmpdir:
+        module = get_module_name(c_file)
+
+        header_file = "{}_datastructure.h".format(module)
```
Comment (tautschnig) [inline scripts/extract_type_header.py]: Nit pick: `header_file = module + "_datastructure.h"` would do.

File: scripts/extract_type_header.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+import shutil
+import subprocess
+import sys
+import textwrap
+from tempfile import TemporaryDirectory
+
+DEFINE_REGEX_HEADER = re.compile(r"\s*#\s*define\s*([\w]+)")
+
+
+def collect_defines(c_file):
```
Comment (tautschnig) [inline scripts/extract_type_header.py:19]: I believe this function would deserve a comment of its own, describing what exactly it is trying to accomplish and where it might get things wrong (as it makes some guesses).

File: scripts/extract_type_header.py [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+
+if __name__ == '__main__':
+    TARGET = None
+    if len(sys.argv) < 3 or len(sys.argv) > 4:
+        print_usage_and_exit()
+    BINARY = sys.argv[1]
+    FILE = sys.argv[2]
+    if len(sys.argv) == 4:
+        TARGET = sys.argv[3]
+    make_header_file(BINARY, FILE, TARGET)
```
Comment (tautschnig) [inline scripts/extract_type_header.py:144]: Random place: shouldn't at least one of the tests also make use of this script? Otherwise it lives here and no-one known when, where, and how to use it.

File: src/goto-instrument/dump_c.cpp [hunk 14]
```diff
@@ -26,33 +27,14 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include <iostream>
 
-// clang-format off
+dump_c_configurationt dump_c_configurationt::default_configuration =
+  dump_c_configurationt().disable_include_headers();
```
Comment (tautschnig) [inline src/goto-instrument/dump_c.cpp]: It seems odd that the "default configuration" is not what the default constructor generates.

File: src/goto-instrument/dump_c.cpp [hunk 15]
```diff
@@ -208,6 +192,22 @@ void dump_ct::operator()(std::ostream &os)
         global_decl_stream << "// " << symbol.name << '\n';
         global_decl_stream << "// " << symbol.location << '\n';
 
+        std::string location_file = id2string(symbol.location.get_file());
+        if(location_file[location_file.length() - 1] == 'h')
```
Comment (tautschnig) [inline src/goto-instrument/dump_c.cpp]: What if `location_file` is empty? And what's the magic behind ending in a `h` - why is that a distinguishing factor?

File: src/goto-instrument/dump_c.cpp [hunk 16]
```diff
@@ -208,6 +192,22 @@ void dump_ct::operator()(std::ostream &os)
...
+        {
+          std::vector<std::string>::iterator it =
+            find(header_files.begin(), header_files.end(), location_file);
+          if(it == header_files.end())
+          {
+            header_files.push_back(location_file);
+            std::vector<char> v;
```
Comment (tautschnig) [inline src/goto-instrument/dump_c.cpp]: Unused?

File: src/goto-instrument/dump_c.cpp [hunk 17]
```diff
@@ -208,6 +192,22 @@ void dump_ct::operator()(std::ostream &os)
...
+
+            global_decl_header_stream
+              << "#include \""
+              << split_string(location_file, '/', true, true).back() << "\"\n";
+          }
+        }
```
Comment (tautschnig) [inline src/goto-instrument/dump_c.cpp:211]: How is this related to the commit message?

File: src/goto-instrument/dump_c.cpp [hunk 18]
```diff
@@ -1466,3 +1478,47 @@ void dump_cpp(
     new_cpp_language);
   out << goto2cpp;
 }
+
+bool module_local_declaration(const symbolt &symbol, const std::string module){
```
Comment (tautschnig) [inline src/goto-instrument/dump_c.cpp]: `static`?

File: src/goto-instrument/dump_c.cpp [hunk 19]
```diff
@@ -1466,3 +1478,47 @@ void dump_cpp(
...
+  std::string filename = id2string(symbol.location.get_file());
+  std::string symbol_module = strip_string(id2string(symbol.module));
```
Comment (tautschnig) [inline src/goto-instrument/dump_c.cpp:1500]: Why is `strip_string` necessary _only_ here?

File: src/goto-instrument/dump_c.cpp [hunk 20]
```diff
@@ -1466,3 +1478,47 @@ void dump_cpp(
...
+  std::vector<std::string> name_components = split_string(filename, '/', true, true);
```
Comment (tautschnig) [inline src/goto-instrument/dump_c.cpp]: I believe you want `get_base_name`?

File: src/goto-instrument/dump_c_class.h [hunk 21]
```diff
@@ -74,6 +86,46 @@ struct dump_c_configurationt final
   /// The config used for dump-c-type-header
   static dump_c_configurationt type_header_configuration;
 
+  dump_c_configurationt disable_include_functions_decl(bool value=false){
```
Comment (tautschnig) [inline src/goto-instrument/dump_c_class.h]: If it's named `disable_...` then I wouldn't expect a parameter that could actually cause the opposite effect.

File: src/util/irep_ids.def [hunk 22]
```diff
@@ -19,6 +19,7 @@ IREP_ID_ONE(x86_extended)
 IREP_ID_TWO(C_source_location, #source_location)
 IREP_ID_TWO(C_end_location, #end_location)
 IREP_ID_TWO(C_is_padding, #is_padding)
+IREP_ID_ONE(C_no_dump)
```
Comment (tautschnig) [inline src/util/irep_ids.def]: `C_do_not_dump` maybe?

**Review by tautschnig — COMMENTED** [review]
I have myself frequently hoped for such a tool, so I much appreciate thus. Just some need to clean things up a bit.

File: scripts/extract_type_header.py (hunk 13 above)
Comment (mmuesly) [reply scripts/extract_type_header.py:144]: I add a test.

File: src/goto-instrument/dump_c.cpp (hunk 20 above)
Comment (mmuesly) [reply src/goto-instrument/dump_c.cpp]: Yes! Thanks, I missed that.

File: src/goto-instrument/dump_c.cpp (hunk 19 above)
Comment (mmuesly) [reply src/goto-instrument/dump_c.cpp:1500]: Where else would you expect it? module is already stripped as it originates from the command line.
Should I re strip it?

File: src/goto-instrument/dump_c.cpp (hunk 17 above)
Comment (mmuesly) [reply src/goto-instrument/dump_c.cpp:211]: I will change the commit message.

The idea is to include the transitive header files that are required to make the new header file compile.

File: scripts/extract_type_header.py (hunk 12 above)
Comment (mmuesly) [reply scripts/extract_type_header.py:19]: I added a comment.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4bdabceeabda4b6fa74b20e79e1180edcf314e8d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115657641

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a46f200055da71ca04e8a70f6e037d230e440fa8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115664621

File: src/goto-instrument/dump_c_class.h (hunk 7 above)
Comment (mmuesly) [reply src/goto-instrument/dump_c_class.h]: The stream controlled by this flag is called global_decls. It is used to declare struct names borrowed from other header files.

File: src/goto-instrument/dump_c.cpp (hunk 15 above)
Comment (mmuesly) [reply src/goto-instrument/dump_c.cpp]: I am just looking for header files and assume that all of them are ending in a `h`.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: b661ae1cae2323d02b371c18e45f2b68f4c20ea1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115828755
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Comment by mmuesly** [comment]
@tautschnig thank you for the review! I addressed all of your comments or followed up on them with requests.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 848bb62ee5fe25c8963d4a5adde1b8fb34c432f0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115829189

**Review by peterschrammel — APPROVED** [review]
lgtm

File: src/goto-instrument/dump_c.cpp [hunk 23]
```diff
@@ -493,8 +549,16 @@ void dump_ct::convert_compound(
   }
 
   typet unresolved_clean=unresolved;
-  typedef_typest::const_iterator td_entry=
-    typedef_types.find(unresolved);
+  typedef_typest::const_iterator td_entry;
+  for(td_entry = typedef_types.begin(); td_entry != typedef_types.end();
+      td_entry++)
```
Comment (romainbrenguier) [inline src/goto-instrument/dump_c.cpp]: use range for

File: src/goto-instrument/dump_c.cpp [hunk 24]
```diff
@@ -1432,3 +1496,56 @@ void dump_cpp(
...
+
+static bool
+module_local_declaration(const symbolt &symbol, const std::string module)
+{
+  std::string base_name =
+    get_base_name(id2string(symbol.location.get_file()), true);
+  std::string symbol_module = strip_string(id2string(symbol.module));
+  if(base_name == module && symbol_module == module)
+  {
+    return true;
```
Comment (romainbrenguier) [inline src/goto-instrument/dump_c.cpp]: `return (base_name == module && symbol_module == module)`

File: src/goto-instrument/dump_c.cpp [hunk 25]
```diff
@@ -1432,3 +1496,56 @@ void dump_cpp(
...
+  const bool use_all_headers,
+  const bool include_harness,
+  const namespacet &ns,
+  const std::string module,
+  std::ostream &out)
+{
+  symbol_tablet symbol_table = ns.get_symbol_table();
+  for(symbol_tablet::iteratort it = symbol_table.begin();
+      it != symbol_table.end();
+      it++)
```
Comment (romainbrenguier) [inline src/goto-instrument/dump_c.cpp:1516]: range for?

File: src/goto-instrument/dump_c_class.h [hunk 26]
```diff
@@ -21,6 +21,120 @@ Author: Daniel Kroening, kroening@kroening.com
...
+
+  dump_c_configurationt()
+    : include_function_decls(true),
+      include_function_bodies(true),
+      include_global_decls(true),
+      include_typedefs(true),
+      include_global_vars(true),
+      include_compounds(true),
+      follow_compounds(true),
+      include_headers(false)
```
Comment (romainbrenguier) [inline src/goto-instrument/dump_c_class.h]: prefer in class initializers to member initializers in constructors https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c48-prefer-in-class-initializers-to-member-initializers-in-constructors-for-constant-initializers

File: src/goto-instrument/dump_c_class.h [hunk 27]
```diff
@@ -21,6 +21,120 @@ Author: Daniel Kroening, kroening@kroening.com
...
+    const bool follow_compounds,
+    const bool include_headers)
+    : include_function_decls(include_function_decls),
+      include_function_bodies(include_function_bodies),
+      include_global_decls(include_global_decls),
+      include_typedefs(include_typedefs),
+      include_global_vars(include_global_vars),
+      include_compounds(include_compounds),
+      follow_compounds(follow_compounds),
+      include_headers(include_headers)
```
Comment (romainbrenguier) [inline src/goto-instrument/dump_c_class.h]: I'm not sure this constructor is useful since it seems clearer to initialize a default one and then set the options as needed.

File: src/goto-instrument/dump_c.cpp (hunk 25 above)
Comment (mmuesly) [reply src/goto-instrument/dump_c.cpp:1516]: I am not aware of telling the ranged for loop how to pick this iterator.
`for(auto it : symbol_table){}` iterates a slightly different data structure.

**Comment by mmuesly** [comment]
@romainbrenguier thank you for your comments. Wherever I was able to do so, I followed your suggestion.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 004cd60d5140443047cb8a72d8fa6d6662848526).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116683609

File: src/goto-instrument/dump_c.cpp (hunk 25 above)
Comment (romainbrenguier) [reply src/goto-instrument/dump_c.cpp:1516]: Strange, it should pick up `.begin()` and `.end()`. At least you should use ++it instead of it++.
